### ISSUE TITLE
Early instances not visible in Wayback

### ISSUE THREAD

**Issue by nicolabingham** [ukwa/w3act#614]
NLW report that some early instances of websites (archived with PANDAS) are not visible in Wayback. 
With the examples below, if you append "index.html" to the Wayback URLs, the earlier instances are returned in some cases, but not necessarily the very early content. This might be because we often put the index file as the primary seed in WCT to get to the content on the server. 

See for example: 

http://www.dic-edwards.com/ and http://www.dic-edwards.com/index.html which show different date ranges in QA Wayback. 
https://www.webarchive.org.uk/act/wayback/*/http://www.dic-edwards.com/index.html has instances from 2004 and 2010 https://www.webarchive.org.uk/act/wayback/*/http://www.dic-edwards.com/ has instances from 2008-2015.

Also: 
http://www.bloc.org.uk/ (archived since 2004) https://www.webarchive.org.uk/act/wayback/*/http://www.bloc.org.uk/index.html (instances from 2014 - 2015) https://www.webarchive.org.uk/act/wayback/*/http://www.bloc.org.uk/ (instances from 2013 - 2019) 

http://www.enlli.org/ (archived since 2004) https://www.webarchive.org.uk/act/wayback/*/http://www.enlli.org/ (instances from 2008 - 2019) https://www.webarchive.org.uk/act/wayback/*/http://www.enlli.org/index.html (instances from 2004 - 2010)

http://www.gweini.org.uk/ (archived since 2006) https://www.webarchive.org.uk/act/wayback/*/http://www.gweini.org.uk/ (instances from 2008 - 2019) https://www.webarchive.org.uk/act/wayback/*/http://www.gweini.org.uk/index.html (instances from 2007 and 2010)

http://www.tracrecord.com/ (archived since 2006) https://www.webarchive.org.uk/act/wayback/*/http://www.tracrecord.com/ (instances from 2008 - 2012) https://www.webarchive.org.uk/act/wayback/*/http://www.tracrecord.com/index.html (instances from 2006 - 2010)

http://wales.gov.uk/  (archived since 2006) https://www.webarchive.org.uk/act/wayback/*/http://wales.gov.uk (instances from 2008 - 2019) https://www.webarchive.org.uk/act/wayback/*/http://wales.gov.uk/index.html (instances from 2005 - 2015)

So, in some cases "index.html" returns earlier instances but in most cases the very early stuff is still missing.

**Comment by GilHoggarth** [comment]
No. / and /index.html are not necessarily the same. The very commonly are but that's just a convention, not a rule. So IMO we shouldn't implement an automatic conversion to the same point.
Some sites choose different filenames for the default file served for /; it may not be index.html and implementing this in code will only cause different problems.

I'd instead suggest that curators define a convention when adding URLs to W3ACT. For example, if the default directory file is wanted, ensure no filename is included in W3ACT. If a particular page is wanted, include that filename in W3ACT.

**Comment by anjackson** [comment]
I agree W3ACT shouldn't do anything differently, but it is not uncommon for such URL collapsing to be done at playback time. We already collapse http+https onto the same item, and I think www/www2/no-www. 

However, looking at the OpenWayback documentation, I don't think it's usual to canonicalise the `/index.html` in this way https://github.com/iipc/openwayback/wiki/Technical-overview#canonicalization-scheme

**Comment by GilHoggarth** [comment]
I guess I focussed on W3ACT as this is the w3act github issues list. But, perhaps you're thinking this should be done by wayback, or perhaps when indexing in the CDX (given you mention the SURT)? I still think this is a dangerous addition for the same reason. Yes this collapsing is very common, it was and probably still is the default directory -> filename configuration in Apache and Nginx, but it still isn't a rule.

Effectively, aren't we trying to implement something here that would be helpful to the viewer but in reality the number of variations (index.html, index.htm, index.asp, index,jsp, index.php, index.cgi, welcome.html, ...) means whatever gets implemented would be an assumption on our behalf and thus shifting our data to our interpretation of our data?

**Comment by ato** [comment]
It's not pretty but we (NLA) handle this issue with pandas crawls at indexing time via outbackcdx's alias records:
https://github.com/nla/bamboo/blob/master/ui/src/bamboo/task/Cdx.java#L100-L111

**Comment by ato** [comment]
> the number of variations (index.html, index.htm, index.asp, index,jsp, index.php, index.cgi, welcome.html, ...) means whatever gets implemented would be an assumption

It depends on how you converted your httrack crawls to warc files but if you didn't undo the link rewriting you shouldn't have to worry about these variations as httrack always uses index.html.

**Comment by ikreymer** [comment]
Yeah, agree with @GilHoggarth that changing canonicalization seems a bit risky and potentially expensive, and there are several
 variations on index.*
I think it'd be safer/easier to try this as a replay lookup change. Since pywb already includes fuzzy matching capabilities, that could be added as another fuzzy rule: If no exact match for `com,example)/`, then search for `com,example)/index.*` closest to date of the original

**Comment by ikreymer** [comment]
Or, even simpler to start,  just do a second exact lookup for `com,example)/index.html` if just considering index.html as its the most common.

**Comment by ato** [comment]
More background: one reason this happens specifically for for pandas crawls (as Nicola identifies) is that httrack writes each page to a separate file file on disk not to warc files. Consequently it must rename / to /index.html because you can't have both a file and a directory with the same name.

Most of the time heritrix crawls will pickup both / and /index.html as separate records if they're linked to. Although it's also possible for / was missed if /index.html was used as the only seed and no internal links used /. That's not uncommon in early websites.

**Comment by nicolabingham** [comment]
Thanks for coming in on this one everyone, this is all really helpful. 
I'll remove this ticket from the bugfix milestone and keep it as a question, as it looks like a policy issue that needs further consideration.
I'll rename the ticket as there might be some genuine missing instances which we can pick up separately.
